### PR TITLE
DDF-6215 Upgrade to Karaf 4.2.9

### PR DESCRIPTION
**Pull request by SmithJosh**
#### What does this PR do?
Upgrades to Karaf 4.2.9, and along with it upgrades a bunch of other dependencies to match the version that Karaf uses.

There was an attempt to upgrade previously, which ran into a number of issues (https://github.com/codice/ddf/issues/5798). I ran into a few of the same issues for this PR. Here's the fixes: 

> Some change, somewhere, has led to a problem with the bnd library we use (stacktrace here). There are two versions that I can see in the system and neither seems to correctly parse some classfile.

This was a problem parsing one of the cxf feature definitions. I copied the fix from https://github.com/apache/cxf/pull/602

> Some change has apparently altered startup order such that when running component tests, many (perhaps all) of our filters cannot be injected because their target servlet contexts have already started.

I changed our DelegateServletFilter to be a Jetty Handler, so it doesn't need to be injected into each individual servlet context. There was no way I could find to access the servlet contexts before they were started. The handler approach accomplishes the same thing - running filters for all requests - but filters must now implement the `org.codice.ddf.platform.filter.http.HttpFilter` interface instead of `javax.servlet.Filter` (Jetty Handlers accept HttpServletRequests instead of ServletRequests).

This PR depends on: https://github.com/codice/thirdparty/pull/97

#### Who is reviewing it? 

@blen-desta @bakejeyner @garrettfreibott @stustison 

#### Select relevant component teams: 

@codice/security 

#### How should this be tested?
Verify:
* Full build with itests passes
* Error pages are still registered. If you refresh the `platform-paxweb-jettyconfig` bundle, you should see a bunch of messages indicating that error pages are being injected as the webapps restart. Also, produce an error in the UI (e.g. 403 by hitting /admin as guest) and verify you see the expected DDF error page.
* HttpFilters are run on all requests and after the SecurityFilters (WebSsoFilter, LoginFilter, etc.). You can turn up the logs for `org.codice.ddf.pax.web.jetty` to TRACE and you should see a message `Delegating to 5 HttpFilters` when you make a request.
* The metrics filter still works. Hit `/metrics` and verify the `ddf_platform_*` metrics look realistic.
* HttpFilters are run in the correct order: SecurityJavaSubjectFilter first, DoPrivilegedFilter last, anything else in between

#### Any background context you want to provide?
This is part of an ongoing effort to reduce the CVE level in DDF

#### What are the relevant tickets?
Fixes: #6215

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-directorymonitor/pom.xml [hunk 1]
```diff
@@ -324,7 +324,7 @@
         <dependency>
             <groupId>ch.qos.cal10n</groupId>
             <artifactId>cal10n-api</artifactId>
-            <version>0.7.4</version>
+            <version>${cal10n.version}</version>
```
[the remark this answers is not in the document]
Comment (SmithJosh) [reply catalog/core/catalog-core-directorymonitor/pom.xml:327]: From http://cal10n.qos.ch/
> Compiler Assisted Localization, abbreviated as CAL10N (pronounced as "calion")

I'm not sure if that's cal + lion
or if it rhymes with scallion 😄

[the remark this answers is not in the document]
Comment (bakejeyner) [reply catalog/core/catalog-core-directorymonitor/pom.xml:327]: LOL lets make a wrap for it and stick it in the comments

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  private final List<Filter> filters;
+
+  public DelegatingSecurityFilter(List<Filter> filters) {
+    this.filters = filters;
+  }
+
+  @Override
+  public void init() {
+    LOGGER.debug("Initialized " + DelegatingSecurityFilter.class.getSimpleName());
```
Comment (bdeining) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java:49]: this should be parameterized

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      new ProxyFilterChain(filters).doFilter(request, response);
+    } catch (ServletException e) {
+      throw new AuthenticationException("Error in global ServletFilter chain", e);
+    }
+    chain.doFilter(request, response);
+  }
+
+  @Override
+  public void destroy() {
+    LOGGER.debug("Destroyed " + DelegatingSecurityFilter.class.getSimpleName());
```
Comment (bdeining) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java:66]: this should be parameterized

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Override
+  public void init() {
+    LOGGER.debug("Initialized " + DelegatingSecurityFilter.class.getSimpleName());
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+      throws IOException, AuthenticationException {
+    LOGGER.debug("Delegating to {} global ServletFilters.", filters.size());
```
Comment (bdeining) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java:55]: the logging in this method might be more appropriate at the trace level

**Comment by SmithJosh** [comment]
> ❗ `FilterInjector` and `DelegateServletFilter` are being used to hook in metrics to all servlets in the system using the `ServletMetrics` filter. I could not find a way to get the Pax Web whiteboard to hook into all servlets in the system which is why I just fixed the injector and delegate.
> https://github.com/codice/ddf/blob/master/platform/metrics/metrics-servlet-filter/src/main/java/org/codice/ddf/metrics/servlet/ServletMetrics.java
> 
> I think we could support the use case with a custom Jetty `Handler`.
> https://www.eclipse.org/jetty/documentation/current/jetty-handlers.html#writing-custom-handlers
> 
> The GZIP handler looks like a good example of connecting to all servlet requests and responses.
> https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java
> https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/config/modules/gzip.mod
> https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/config/etc/jetty-gzip.xml
> 
> Looks like there is even a way to register handlers directly as OSGi services which would be easier then trying to figure out how to get a mod to work with Pax Web.
> http://ops4j.github.io/pax/web/SNAPSHOT/User-Guide.html#using-handler-and-connectors-as-services

@pklinef I had the same issue trying to get a whiteboard service to hook into all servlets. Pax Web doesn't support it. That's why I ended up with the solution in this PR. Services implementing the SecurityFilter interface do get applied to all requests, so I made the `DelegateServletFilter` into a `DelegateSecurityFilter`. It's a workaround, but it performs the intended behavior. And it appears that the Pax Web bug preventing us from hooking into all servlets via the whiteboard will be fixed in pax-web 8, so I thought that was good enough for now.

I can look into the Handler suggestion, but I'm a little unclear what the issue with this solution is

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegateServletFilter.java [hunk 5]
```diff
@@ -1,54 +0,0 @@
...
-import javax.servlet.Filter;
-import javax.servlet.FilterChain;
-import javax.servlet.FilterConfig;
-import javax.servlet.ServletException;
-import javax.servlet.ServletRequest;
-import javax.servlet.ServletResponse;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-public class DelegateServletFilter implements Filter {
```
Comment (Lambeaux) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegateServletFilter.java:27]: ❓ Isn't this going to break metrics and other cross-cutting concerns? I think we have some dependents on this class without any viable substitutes as of yet.

Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegateServletFilter.java:27]: > Isn't this deletion going to break metrics and other cross-cutting concerns (i.e. client info filter but I doubt anyone still cares about that)? I think we have some dependents on this class without any viable substitutes as of yet.

I only changed the name of this class. It's now `DelegateSecurityFilter`, but it still performs the same behavior, picking up all registered ServletFilters and passing incoming requests off to them.

And what do you mean by "dependencies"? I wasn't aware anybody used this class itself. They should just be registering ServletFilters which will get picked up by this.  

> I don't think your SecurityFilter workaround is valid in all current cases.

Do you have an example by chance where this wouldn't work?

**Comment by SmithJosh** [comment]
@pklinef I looked at the `ServletMetrics` filter and I think I understand your concern now. So since the delegate filter is now part of the SecurityFilter chain, it doesn't have access to the ServletFilter chain in the actual webapp that gets hit like it did before and can't accurately measure request latency. And indeed, I was able to confirm the latency metrics are wrong.

I'm looking into the Jetty `Handler` idea to address that. It sounds like your concern doesn't apply to the rest of the filters, so I'll just plan on changing the ServletMetrics filter for now.

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java (hunk 4 above)
Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingSecurityFilter.java]: I changed it to trace

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * Hands the request off to a set chain of servlet {@link Filter}s. Since SecurityFilters are run on
+ * each request, this provides a mechanism to add global servlet filters. As of OSGi R6, there is a
+ * proper way to define global servlets/filters/listeners/etc., defined by the HTTP Whiteboard spec.
+ * However, pax-web does not yet implement that feature, so we're left using this workaround.
+ *
+ * <p>When https://ops4j1.jira.com/browse/PAXWEB-1123 is resolved, this workaround should be
+ * revisited.
+ */
+public class DelegatingHttpFilterHandler extends HandlerWrapper {
```
Comment (SmithJosh) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java:41]: The delegating servlet filter has a new home 😂 This time it's the Jetty Handler chain

File: platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * An HttpFilter is an object that performs filtering tasks on either the request to a resource (a
+ * servlet or static content), or on the response from a resource, or both.
+ *
+ * <p>It is modelled after {@link javax.servlet.Filter} but its {@code doFilter} method accepts
+ * {@link HttpServletRequest} and {@link HttpServletResponse} arguments instead.
+ */
+public interface HttpFilter {
+
+  void doFilter(
```
Comment (SmithJosh) [inline platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java:30]: The Jetty Handlers take HttpServletRequests. In order to inject stuff into that chain, I needed to have a filter interface that takes HttpServletRequests as well. Oddly enough this is not already a thing.

Anything filters we used to inject via the delegate filter will need to be changed to implement this new interface. That will require changing the method signature and removing the init/destroy methods since those aren't used anymore.

File: platform/metrics/metrics-servlet-filter/src/main/java/org/codice/ddf/metrics/servlet/ServletMetrics.java [hunk 8]
```diff
@@ -18,45 +18,34 @@
...
-import javax.servlet.ServletResponse;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import org.codice.ddf.platform.filter.HttpFilter;
+import org.codice.ddf.platform.filter.HttpFilterChain;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class ServletMetrics implements Filter {
+public class ServletMetrics implements HttpFilter {
```
Comment (SmithJosh) [inline platform/metrics/metrics-servlet-filter/src/main/java/org/codice/ddf/metrics/servlet/ServletMetrics.java:29]: Here's an example of converting an existing Filter into an HttpFilter

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    LOGGER.trace("Delegating to {} HttpFilters.", filters.length);
+
+    ProxyHttpFilterChain filterChain =
+        new ProxyHttpFilterChain(filters, getHandler(), target, baseRequest);
+    filterChain.doFilter(request, response);
+  }
+
+  private HttpFilter[] getFilters() {
+    HttpFilter[] filters = new HttpFilter[filterTracker.size()];
+    return filterTracker.getServices(filters);
```
Comment (SmithJosh) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java:75]: This is getting injected into the Jetty Handler chain via the jetty.xml config, so I couldn't just inject a service reference list like it did before. If anybody has better ideas for how to do this, I'm open to suggestions.

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegateServletFilter.java (hunk 5 above)
Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegateServletFilter.java:27]: You were right. The DoPrivilegedFilter wasn't working either, since the doPrivileged call terminated as soon as all the Filters had run. The delegate filter wasn't really running the the filters as part of the chain.

I believe I fixed that now, so marking this as resolved

File: platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 10]
```diff
@@ -15,43 +15,24 @@
...
-        </argument>
-        <argument ref="executor"/>
-    </bean>
-
-    <bean id="securityJavaSubjectThreadFactory"
-          class="org.codice.ddf.platform.util.StandardThreadFactoryBuilder"
-          factory-method="newThreadFactory">
-        <argument value="securityJavaSubjectThread"/>
-    </bean>
+    <service id="securityJavaSubjectFilter" interface="org.codice.ddf.platform.filter.HttpFilter">
```
Comment (pklinef) [inline platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: ❓ Did we lose the ability to make sure these core handlers are executed in the proper order?  I guess we could use service ranking if we need it again.

File: platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * An HttpFilter is an object that performs filtering tasks on either the request to a resource (a
+ * servlet or static content), or on the response from a resource, or both.
+ *
+ * <p>It is modelled after {@link javax.servlet.Filter} but its {@code doFilter} method accepts
+ * {@link HttpServletRequest} and {@link HttpServletResponse} arguments instead.
+ */
+public interface HttpFilter {
```
Comment (pklinef) [inline platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java:28]: ❓ Why is this part of `security-filter-api` instead of part of `platform-paxweb-jettyconfig` or some other Jetty or HTTP concerned bundle?  This interface does not seem related to security as it is.

File: platform/metrics/metrics-servlet-filter/pom.xml [hunk 12]
```diff
@@ -44,6 +44,12 @@
             <artifactId>slf4j-api</artifactId>
             <version>${org.slf4j.version}</version>
         </dependency>
+        <dependency>
+            <groupId>ddf.platform</groupId>
+            <artifactId>platform-paxweb-jettyconfig</artifactId>
```
Comment (pklinef) [inline platform/metrics/metrics-servlet-filter/pom.xml:49]: ❓ Is this dependency actually needed here?  It should only depend on the API bundle with the `HttpFilter` interface.

File: platform/platform-paxweb-jettyconfig/pom.xml [hunk 13]
```diff
@@ -103,6 +103,13 @@
             <artifactId>security-core-impl</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-core</artifactId>
```
[the remark this answers is not in the document]
Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/pom.xml:108]: Oops, I forgot to clean this up. It shouldn't be here.

File: platform/metrics/metrics-servlet-filter/pom.xml (hunk 12 above)
Comment (SmithJosh) [reply platform/metrics/metrics-servlet-filter/pom.xml:49]: You're right. I'd originally made the new interface in the paxweb-jettyconfig bundle and forgot to clean this up.

File: platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java (hunk 11 above)
Comment (SmithJosh) [reply platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java:28]: Yeah I guess it's not necessarily related to security. I put it here since most of the filter implementations are security-related, but it may make more sense in a new api bundle.

I don't think it should go into `platform-paxweb-jettyconfig` since that would pull in a bunch of other transitive dependencies as well.

File: platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 10 above)
Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: Hm... I'm not sure what guarantees we have about execution order anymore. If we need to preserve the existing order (ClientInfoFilter, ResponseFilter, SecurityJavaSubjectFilter, ServletMetrics, DoPrivilegedFilter), I'll look into it

File: platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java (hunk 11 above)
Comment (pklinef) [reply platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java:28]: A new API bundle would be safest.  Currently `platform-api` looks pretty safe but that runs the risk of being exposed to other unrelated dependencies in the future.  Already it depends on `javax.validation`.

Comment (SmithJosh) [reply platform/security-filter-api/src/main/java/org/codice/ddf/platform/filter/HttpFilter.java:28]: Ok I'll create a new bundle

File: platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 10 above)
Comment (pklinef) [reply platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: Originally I was intentional abut making sure the filter delegate went after the security java subject filter so that the delegated filters could assume a subject would be available if the user was logged in.  I don't think the current filters care about that currently.  It is probably ok how you have it currently.

Comment (pklinef) [reply platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: Though I wonder if it will impact the Java security manager policy.  Before we were able to keep policies details from leaking up to global filters.

Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: Good point. I'll see what I can do to make the DoPrivilegedFilter run last. Not sure if the ServiceTracker supports service ranking as I'm using it, but that would be easiest

Comment (pklinef) [reply platform/platform-paxweb-jettyconfig/src/main/resources/OSGI-INF/blueprint/blueprint.xml:30]: You could do something similar to our `SortedServiceList`.

https://github.com/codice/ddf/blob/master/platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/SortedServiceList.java#L57-L64

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/ProxyHttpFilterChain.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(ProxyHttpFilterChain.class);
+
+  private final Iterator<HttpFilter> iterator;
+  private final Handler handler;
+  private final String target;
+  private final Request baseRequest;
+
+  public ProxyHttpFilterChain(
+      List<HttpFilter> filters, Handler handler, String target, Request baseRequest) {
+    this.iterator = filters.iterator();
```
Comment (Lambeaux) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/ProxyHttpFilterChain.java:46]: ❓ Iterators are stateful. Does a new instance of this chain get created for each request? If not, and all requests share it, then how is this working?

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+  @Override
+  public void handle(
+      String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
+      throws IOException, ServletException {
+    LOGGER.trace("Delegating to {} HttpFilters.", httpFilters.size());
+
+    ProxyHttpFilterChain filterChain =
+        new ProxyHttpFilterChain(httpFilters, getHandler(), target, baseRequest);
+    filterChain.doFilter(request, response);
```
Comment (Lambeaux) [inline platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/DelegatingHttpFilterHandler.java:101]: 👍  I see - should be fine

File: platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/ProxyHttpFilterChain.java (hunk 14 above)
Comment (SmithJosh) [reply platform/platform-paxweb-jettyconfig/src/main/java/org/codice/ddf/pax/web/jetty/ProxyHttpFilterChain.java:46]: Yep, the ProxyHttpFilterChain is single-use. A new instance gets created for each request

**Comment by bakejeyner** [comment]
Hero ✔️ 
- verified filters are re-injected after refreshing bundle
- error page is displayed
- filters are being injected, and the default http filters are being ran
- metrics look fine
- security filters run in correct order

**Comment by SmithJosh** [comment]
build now
